### ISSUE TITLE
BBT fails to initialize when Zotero translators are missing

### ISSUE THREAD

**Issue by schferbe** [retorquere/zotero-better-bibtex#1313]
Cross posting from https://forums.zotero.org/discussion/comment/342128.

I am using Better BibTeX 5.1.148 with Zotero 5.0.76 and recently it hangs when starting. Zotero uses one CPU core completely. I've tried to reset the translators and got an error that I reported with ID 1210104811.

Most parts of Zotero work but are really slow. Others don't work e.g. indexing of imported files. 

**Report ID:** AULY3SKZ-euc

**Expected behavior:** Initialization finishes in finite time.

**Actual behavior:** Initialization carries on indefinitely.

**Comment by schferbe** [comment]
LHWXN6X7-euc

I tried playing around a bit. It's a pain though since everything is so slow.

**Comment by retorquere** [comment]
Can you try resetting the translators in the Zotero preferences and restarting Zotero? I see errors in your log that the Blogger translator is missing, and for some reason this breaks translators init, and that in turn prevents BBT from starting up completely.

**Comment by schferbe** [comment]
I have tried but that gives an error with the following report:
```
[JavaScript Error: "Invalid or missing translator metadata JSON object in Japan Times Online.js" {file: "chrome://zotero/content/xpcom/translation/translators.js" line: 273}]

[JavaScript Error: "{better-bibtex! +null} Better BibTeX load failed {"operation":"stat","path":"/home/ben/Picloud/Documents/Zotero/translators/EurasiaNet.js","unixErrno":2,"fileName":"(unknown module)","lineNumber":0} "]

version => 5.0.76, platform => Linux x86_64, oscpu => Linux x86_64, locale => en-US, appName => Zotero, appVersion => 5.0.76, extensions => Zotero LibreOffice Integration (5.0.20.SA.5.0.76, extension), Better BibTex for Zotero (5.1.148, extension), Shockwave Flash (32.0.0.270, plugin)
```

**Comment by retorquere** [comment]
Try uninstalling BBT, restarting, resetting translators, and reinstalling BBT

**Comment by schferbe** [comment]
I uninstalled BBT and tried to reset translators. Following error occurred:
```
[JavaScript Error: "Unix error 2 during operation stat on file /home/ben/Picloud/Documents/Zotero/translators/Google Presentation.js (Datei oder Verzeichnis nicht gefunden)" {file: "(unknown module)" line: 0}]

[JavaScript Error: "Unix error 2 during operation stat on file /home/ben/Picloud/Documents/Zotero/translators/Google Presentation.js (Datei oder Verzeichnis nicht gefunden)" {file: "(unknown module)" line: 0}]

[JavaScript Error: "Unix error 2 during operation stat on file /home/ben/Picloud/Documents/Zotero/translators/Google Presentation.js (Datei oder Verzeichnis nicht gefunden)" {file: "(unknown module)" line: 0}]

[JavaScript Error: "Unix error 2 during operation stat on file /home/ben/Picloud/Documents/Zotero/translators/Google Presentation.js (Datei oder Verzeichnis nicht gefunden)" {file: "(unknown module)" line: 0}]

version => 5.0.76, platform => Linux x86_64, oscpu => Linux x86_64, locale => en-US, appName => Zotero, appVersion => 5.0.76, extensions => Zotero LibreOffice Integration (5.0.20.SA.5.0.76, extension), Shockwave Flash (32.0.0.270, plugin)
```
Afterwards I installed BBT 5.1.151 from file. Then the same problem as before could be observed.

Is this even a problem with BBT? Should I reinstall Zotero?

**Comment by retorquere** [comment]
The missing translator errors should be resolved first; in the procedure I outlined before I had assumed that doing this without BBT installed would allow it to be fixed. If "reset translators" still gives errors without BBT installed, that's clearly a Zotero error. That needs to be resolved first, and then we can see if there's still a BBT problem, but there may not be. BBT is only partially started when the error you show happens, the startup of BBT is interrupted by it, and after that I can imagine a lot goes wrong as the parts that did start try to use the parts that didn't, causing errors, which need to be handled, which takes time, etc.

You could try installing Zotero, but I reckon the Zotero team will be interested in learning what went wrong so they can prevent it happening to others. If you can reproduce a problem without BBT installed, I guarantee they'll want to know about it.

**Comment by retorquere** [comment]
What's `Picloud` BTW?

**Comment by schferbe** [comment]
Alright, I'll continue posting in the Zotero forum then. Thanks for your help. 

`Picloud` is just the folder I save my lib in. Which is synced over Nextcloud.

**Comment by retorquere** [comment]
Are your `sqlite` databases in there? That's the first thing they're going to tell you to stop doing. Cloud sync of the zotero databases is absolutely not safe and is very likely the cause of your problems (https://www.zotero.org/support/sync, section "Alternative Syncing Solutions"). Your database being corrupt because of cloud sync would explain a lot of what you're seeing.

**Comment by schferbe** [comment]
Thank you for the link. I didn't know that. Now I changed the syncing. 

Then I tried updating the translators and styles and resetting them without BBT installed. Then installed BBT. However this does not resolve the problem. Do you have any ideas that I can try?

When I try resetting with BBT installed I get this error. Without BBT I don't get an error.
```
[JavaScript Error: "{better-bibtex! +null} Better BibTeX load failed {"operation":"stat","path":"/home/ben/Zotero/translators/Citavi 5 XML.js","unixErrno":2,"fileName":"(unknown module)","lineNumber":0} "]

version => 5.0.76, platform => Linux x86_64, oscpu => Linux x86_64, locale => en-US, appName => Zotero, appVersion => 5.0.76, extensions => Zotero LibreOffice Integration (5.0.20.SA.5.0.76, extension), Better BibTex for Zotero (5.1.151, extension), Shockwave Flash (32.0.0.270, plugin)
```

**Comment by retorquere** [comment]
That depends on whether you mean the error doesn't happen without BBT installed, or just that you don't see it. You'll need to run your database through the zotero repair tool - the current state of your database is likely still corrupted, and I can't tell how much of that damage affects BBT.

**Comment by retorquere** [comment]
To be clear, just changing the syncing doesn't undo the damage. The damage must be verified to be fixed. There's no db fixer for the bbt database, if you want I can give it a go if you get me a copy (taken when zotero **not** running, cmd-q if you're on Mac), otherwise you'd need to delete the bbt database and let bbt regenerate the keys.

**Comment by schferbe** [comment]
I used Zoteros dbfix on my sqlite which generated a database with a smaller file size. However Zotero never said it was corrupted. Replacing the Zotero-DB -> still no remedy. Then I deleted the BBT sqlite. Also didn't help.

I'd guess that the error doesn't occur if BBT is not installed. I generated a debug log without BBT where I update the styles and translators and reset them. ID: `D1532172124`  Do you have access to that?

**Comment by retorquere** [comment]
If you'd reset translators without BBT installed, do you have the citavi 5 xml translator in your translators folder?

**Comment by retorquere** [comment]
I don't have access to zotero debug logs, only my own

**Comment by retorquere** [comment]
Can I get a copy of your db (taken when zotero is not running)?

**Comment by schferbe** [comment]
`Citavi 5 XML.js` is present in the translators folder.

[debug.log](https://github.com/retorquere/zotero-better-bibtex/files/3788699/debug.log)

**Comment by retorquere** [comment]
(I don't think the db fixer says it's corrupted, it just tries to save what it can)

**Comment by schferbe** [comment]
Can I send you a private message with the database?

**Comment by schferbe** [comment]
They look alright. Do they need the execution bit set? (I'm on Linux ext4)

**Comment by retorquere** [comment]
Can you start Zotero with the `-P` flag, create a new profile, and click `Start Zotero`? You'll start with a new profile but your existing data directory. If BBT installs in that clean profile normally, we'd know whether the problem is triggered by Zotero or by your profile. Yuou can later delete the profile by starting with `-P` again.

**Comment by schferbe** [comment]
Zotero starts in another data directory with a new profile. When I install BBT everything is alright. Then I copied my old data directory into the new location and started Zotero again. The problem is there again. Also when I use the clean/working new data directory as start and only copy my storage folder and sqlite-DB into the new data directory the problem is there.

So I guess it has to do with my storage or my database and not with the profile.

Probably I'll only get to answer you the next time on Friday. But thank you very much for your help and very quick answers so far.

**Comment by averter** [comment]
I am experiencing the same problem as @schferbe

**Comment by retorquere** [comment]
I'll need a debug log

**Comment by averter** [comment]
on it, ID is D152348727

**Comment by retorquere** [comment]
I can't see those, sorry, I should have been clearer; I need a bbt debug log. Restart with debugging enabled, then pick the "bbt debug log" option from the help menu.

**Comment by retorquere** [comment]
(bbt debug log id's end in euc or apse)

**Comment by averter** [comment]
Apologies, I haven't used the debug options from better bibtex for quite a while. ID is M6LWQECZ-euc

**Comment by retorquere** [comment]
@averter I don't see any problems with translators in that log. Can you open a separate issue?

**Comment by retorquere** [comment]
(the problem isn't just that it's on linux. My test suite runs on linux and all tests would have broken over the translators not initializing)

**Comment by averter** [comment]
What should I call the new issue? The diagnosis is also just as @schferbe described: better bibtex hanging, making zotero slow and completely taking over one CPU

**Comment by retorquere** [comment]
You can just call it that -- the slowness @schferbe experiences is more likely (by current diagnosis at least) because BBT fails to initialize fully. I don't see evidence of that in your log.

In the new issue, can you click around a bit before submitting the log? I pretty much only see BBT starting and then nothing.

**Comment by schferbe** [comment]
@retorquere I send you an e-mail with a link to my data directory.

**Comment by retorquere** [comment]
Zotero/BBT starts without a problem on that DB/storage folder for me, so it's not just that.

Can you try 5055? It'll be pretty slow because it logs a *lot*, but maybe it will give me some insight on what's going on.

**Comment by retorquere** [comment]
Closing for inactivity.

**Comment by schferbe** [comment]
Sorry for the inactivity. My problem seems to be fixed with v5.1.160. Maybe because of #1314 ?

**Comment by retorquere** [comment]
A few bugs have been fixed in the interim, so I don't know which of these issues affected you, but good to hear it's resolved.
